tests/utilities/test_auto_restart.py

```diff
@@ -1060,3 +1062,142 @@ def run(should_fail, resume):
+            print(message)
+            os.kill(os.getpid(), signal.SIGUSR1)
+            # small optimization to skip as soon as async signal as being triggered.
+            t0 = time()
+            while not self.trainer._terminate_gracefully:
+                t1 = time()
+                if (t1 - t0) > 0.1:
+                    self.trainer._terminate_gracefully = True
+
+    def training_step(self, batch, batch_idx):
+        self.seen_train_batches.append(batch)
+        should_signal = self.trainer.fit_loop.epoch_loop._is_training_done if self.on_last_batch else batch_idx == 2
+        if self.failure_on_step and self.failure_on_training and should_signal:
+            self._signal("training_step")
+        return super().training_step(batch, batch_idx)
+
+    def validation_step(self, batch, batch_idx):
+        should_signal = (
+            self.trainer.fit_loop.epoch_loop.val_loop.epoch_loop.batch_progress.is_last_batch
+            if self.on_last_batch
+            else batch_idx == 2
+        )
+        if self.failure_on_step and not self.failure_on_training and should_signal:
+            self._signal("validation_step")
+        return super().validation_step(batch, batch_idx)
+
+    def training_epoch_end(self, outputs) -> None:
+        if not self.failure_on_step and self.failure_on_training:
+            self._signal("training_epoch_end")
+
+    def validation_epoch_end(self, outputs) -> None:
+        if not self.failure_on_step and not self.failure_on_training:
+            self._signal("validation_epoch_end")
+
+    def train_dataloader(self):
+        return DataLoader(RandomDataset(32, 4))
+
+    def val_dataloader(self):
+        return DataLoader(RandomDataset(32, 4))
+
+
+def _fit_model(
+    tmpdir, should_signal, val_check_interval, failure_on_step, failure_on_training, on_last_batch, status=None
+):
+    seed_everything(42)
+    model = TestAutoRestartModelUnderSignal(should_signal, failure_on_step, failure_on_training, on_last_batch)
+
+    trainer_kwargs = dict(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        limit_train_batches=4,
+        limit_val_batches=4,
+        val_check_interval=val_check_interval,
+        num_sanity_val_steps=0,
+    )
+
+    trainer = Trainer(**trainer_kwargs)
+    if should_signal:
+        with pytest.raises(ExitGracefullyException, match=status):
+            trainer.fit(model)
+    else:
+        trainer.fit(model)
+    assert trainer._terminate_gracefully == should_signal
+
+    return model
+
+
+@pytest.mark.parametrize("on_last_batch", [False, True])
+@pytest.mark.parametrize("val_check_interval", [0.5, 1.0])
+@pytest.mark.parametrize("failure_on_training", [False, True])
+@pytest.mark.parametrize("failure_on_step", [False, True])
+@mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": "1"})
+@RunIf(min_torch="1.7.0", skip_windows=True)
+def test_auto_restart_under_signal(on_last_batch, val_check_interval, failure_on_training, failure_on_step, tmpdir):
+
```

Comment:
the docstring is not attached to the function! 
black malfunction?

Reply:
Done !